gitindex/index.go

```diff
@@ -532,6 +518,225 @@ func newIgnoreMatcher(tree *object.Tree) (*ignore.Matcher, error) {
 	return ignore.ParseIgnoreFile(strings.NewReader(content))
 }
 
+func prepareDeltaBuild(options Options, repository *git.Repository) (repos map[fileKey]BlobLocation, branchMap map[fileKey][]string, branchVersions map[string]map[string]plumbing.Hash, changedOrDeletedPaths []string, err error) {
+	// discover what commits we indexed during our last build
+	shards := options.BuildOptions.FindAllShards()
+	if len(shards) == 0 {
+		return nil, nil, nil, nil, fmt.Errorf("no existing shards found for repository")
+	}
+
+	shard := shards[0]
+	repositories, _, err := zoekt.ReadMetadataPathAlive(shard)
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("reading reposistory metadata from shard %q: %w", shard, err)
+	}
+
+	var existingRepository *zoekt.Repository
+
+	for _, r := range repositories {
+		if r.ID == options.BuildOptions.RepositoryDescription.ID {
+			existingRepository = r
+			break
+		}
+	}
+
+	if existingRepository == nil {
+		return nil, nil, nil, nil, fmt.Errorf("shard %q doesn't contain repository", shard)
+	}
+
+	// Check to see if the branch set is consistent with what we last indexed.
+	// If it isn't consistent, that we can't proceed with a delta build (and the caller should fall back to a
+	// normal one).
+
+	var existingBranchNames []string
+	for _, b := range existingRepository.Branches {
+		existingBranchNames = append(existingBranchNames, b.Name)
+	}
+
+	sort.Strings(existingBranchNames)
+	sort.Strings(options.Branches)
```

Comment:
sorting options.Branches is dangerous, you should be able to rely on the order. This is the same bug around HEAD not being first.

Reply:
You are correct. This is code that was written before we discovered the bug last week, and I'll update this before I mark this as a non-draft PR.